torch/nn/parameter.py

```diff
@@ -196,7 +196,78 @@ def __deepcopy__(self, memo):
             memo[id(self)] = result
             return result
 
-class UninitializedBuffer(UninitializedTensorMixin, torch.Tensor):
+# Metaclass to combine _TensorMeta and the instance check override for Buffer.
+class _BufferMeta(torch._C._TensorMeta):
+    # Make `isinstance(t, Buffer)` return True for custom tensor instances that have the _is_buffer flag.
+    def __instancecheck__(self, instance):
+        return super().__instancecheck__(instance) or (
+            isinstance(instance, torch.Tensor) and getattr(instance, '_is_buffer', False))
+
+
+class Buffer(torch.Tensor, metaclass=_BufferMeta):
+    r"""A kind of Tensor that is to be considered a module parameter but not a model
+    parameter. For example, BatchNorm's ``running_mean`` is not a parameter, but is part of the module's state.
+
+    Buffers are :class:`~torch.Tensor` subclasses, that have a
+    very special property when used with :class:`Module` s - when they're
+    assigned as Module attributes they are automatically added to the list of
+    its buffers, and will appear e.g. in :meth:`~Module.buffers` iterator.
+    Assigning a Tensor doesn't have such effect. One can still assign a Tensor as explicitly by using
+    a the modules `~register_buffer` function.
+
+    Args:
+        data (Tensor): parameter tensor.
+    """
+    def __new__(cls, data=None, requires_grad=False, persistent=True):
+        if data is None:
+            data = torch.empty(0)
+        if type(data) is torch.Tensor or type(data) is Buffer:
+            # For ease of BC maintenance, keep this path for standard Tensor.
+            # Eventually (tm), we should change the behavior for standard Tensor to match.
+            ret = torch.Tensor._make_subclass(cls, data, require_grad=requires_grad)
+            ret.persistent = persistent
+            return ret
+
```

Comment:
Yes, ideally we should update all instances of `type(t) is Parameter` to be `isinstance(t, Parameter)` within PyTorch core. I think this will have to happen for Dynamo to support tensor subclasses.

Reply:
There's something I don't understand here and admittedly didn't understand why this code was needed for `Parameter` so thanks for raising this issue. In the python documentation for the `type` builtin it says:

> With one argument, return the type of an object. The return value is a type object and generally the same object as returned by [object.\_\_class\_\_](https://docs.python.org/3/library/stdtypes.html#instance.__class__).

So I guess I don't get why `type(p) is torch.nn.parameter.Parameter` check would ever fail if p was indeed a Parameter object and also seems like it makes sense it should fail if it is a different class as per the python documentation but I admit some Python ignorance on my part so maybe I'm just misunderstanding something basic here. 

I tried running this example on pytorch 2.0 to try understand this:
```
>>> a = torch.empty(0)
>>> b = torch.nn.Parameter(a)
>>> c = torch.nn.Parameter(b)
>>> print_helper = lambda x: print(f'Type: {type(x)}, isTensor: {isinstance(x, torch.Tensor)}, isParameter: {isinstance(x, torch.nn.Parameter)}')
>>> print_helper(a)
Type: <class 'torch.Tensor'>, isTensor: True, isParameter: False
>>> print_helper(b)
Type: <class 'torch.nn.parameter.Parameter'>, isTensor: True, isParameter: True
>>> print_helper(c)
Type: <class 'torch.nn.parameter.Parameter'>, isTensor: True, isParameter: True
```

I then changed the code `if type(data) is torch.Tensor or type(data) is Parameter` to `if isinstance(data, Parameter):` and I get the following:

```
>>> a = torch.empty(0)
>>> b = torch.nn.Parameter(a)
>>> c = torch.nn.Parameter(b)
>>> print_helper = lambda x: print(f'Type: {type(x)}, isTensor: {isinstance(x, torch.Tensor)}, isParameter: {isinstance(x, torch.nn.Parameter)}')
>>> print_helper(a)
Type: <class 'torch.Tensor'>, isTensor: True, isParameter: False
>>> print_helper(b)
Type: <class 'torch.Tensor'>, isTensor: True, isParameter: True
>>> print_helper(c)
Type: <class 'torch.nn.parameter.Parameter'>, isTensor: True, isParameter: True
```
Which is what I expect since for `b` the check `isinstance(data, Parameter)` would be `False` so I don't understand what you mean here: 

> It is worth noting that for the existing parameter class type(p) is torch.nn.parameter.Parameter returns True when the data passed to the torch.nn.Parameter was a regular torch.Tensor, but does not return True when the data passed was a tensor subclass. On the other hand isinstance(p, torch.nn.parameter.Parameter) will return True for both

Especially the last sentence unless you were referring to subclasses of `torch.nn.Parameter` and not `torch.Tensor`

Reply:
Or is that they `type(torch.nn.Parameter)` should ideally always return `torch.Tensor` but still have `isinstance(p, torch.nn.Parameter)` still work in which case I think I get it now.

Reply:
Hey, running this script should elucidate my comment here! :)

```
import torch
from torch.testing._internal.logging_tensor import LoggingTensor
t = torch.randn(3, 5)
# LoggingTensor is a tensor subclass :)
t_subclass = LoggingTensor(t)
t_param = torch.nn.Parameter(t)
t_subclass_param = torch.nn.Parameter(t_subclass)
print(f"type(t_param) is torch.nn.Parameter = {type(t_param) is torch.nn.Parameter}")
print(f"type(t_subclass_param) is torch.nn.Parameter = {type(t_subclass_param) is torch.nn.Parameter}")
print(f"isinstance(t_param, torch.nn.Parameter) = {isinstance(t_param, torch.nn.Parameter)}")
print(f"isinstance(t_subclass_param, torch.nn.Parameter) = {isinstance(t_subclass_param, torch.nn.Parameter)}")
```

Reply:
Thank you for the clarification! I understand the issue now.